src/librustdoc/html/static/js/search.js

```diff
@@ -167,33 +546,143 @@ window.initSearch = function(rawSearchIndex) {
     }
 
     /**
-     * Executes the query and returns a list of results for each results tab.
-     * @param  {Object}        query          - The user query
-     * @param  {Array<string>} searchWords    - The list of search words to query against
-     * @param  {string}        [filterCrates] - Crate to search in
-     * @return {{
-     *   in_args: Array<?>,
-     *   returned: Array<?>,
-     *   others: Array<?>
-     * }}
+     * Parses the query.
+     *
+     * The supported syntax by this parser is as follow:
+     *
+     * ident = *(ALPHA / DIGIT / "_")
+     * path = ident *(DOUBLE-COLON ident)
+     * arg = path [generics]
+     * arg-without-generic = path
+     * type-sep = COMMA/WS *(COMMA/WS)
+     * nonempty-arg-list = *(type-sep) arg *(type-sep arg) *(type-sep)
+     * nonempty-arg-list-without-generics = *(type-sep) arg-without-generic
+     *                                      *(type-sep arg-without-generic) *(type-sep)
+     * generics = OPEN-ANGLE-BRACKET [ nonempty-arg-list-without-generics ] *(type-sep)
+     *            CLOSE-ANGLE-BRACKET/EOF
+     * return-args = RETURN-ARROW *(type-sep) nonempty-arg-list
+     *
+     * exact-search = [type-filter *WS COLON] [ RETURN-ARROW ] *WS QUOTE ident QUOTE [ generics ]
```

Comment:
Another weird corner case: `->"-"`

rustdoc-search-ref says it's an error, and if hyphen is not a valid character in, then it would be an error.

search.js accepts it, producing this parse result:

```json
{
    "original": "->\"-\"",
    "userQuery": "->\"-\"",
    "typeFilter": -1,
    "elems": [],
    "returned": [
        {
            "name": "-",
            "fullPath": [
                "-"
            ],
            "pathWithoutLast": [],
            "pathLast": "-",
            "generics": []
        }
    ],
    "foundElems": 1,
    "literalSearch": true,
    "error": null
}
```

Reply:
Great catch!